### PR TITLE
encode json serailization of Document and html content to utf-8

### PR DESCRIPTION
**Pull request by nicoxx**
skip url if there is ambiguity in the encoding
use 'requests' instead of urllib2 as it manages encoding better, use simplejson backend for jsonpickle for the same reason

### CODE REVIEW

File: pylintrc_tests [hunk 1]
```diff
@@ -362,7 +362,7 @@ max-args=7
 ignored-argument-names=_.*
 
 # Maximum number of locals for function / method body
-max-locals=15
+max-locals=20
```
Comment (nicoxx) [inline pylintrc_tests]: continue to relax pylint constraints for unit tests

File: requirements.txt [hunk 2]
```diff
@@ -3,4 +3,6 @@ praw
 pyopenssl
 ndg-httpsclient
 pyasn1
-urllib3
\ No newline at end of file
+urllib3
+simplejson
+requests
```
Comment (nicoxx) [inline requirements.txt]: requests instead of urlllib2, simplejson instead of built in module json

File: src/scraper/run_scraper.py [hunk 3]
```diff
@@ -0,0 +1,8 @@
+from scraper.scraper import scrap_and_dump
+import logging
+
+
+
+logging.basicConfig(filename='run_scraper.log')
+scrap_and_dump('recorded_html_02_08')
+
```
Comment (nicoxx) [inline src/scraper/run_scraper.py]: to run easily the scraper...

File: src/scraper/scraper/scraper.py [hunk 4]
```diff
@@ -40,34 +61,43 @@ def _get_invalid_regex():
...
     documents = starmap(Document, filtered_links_html)
     docs_as_jsons = imap(jsonpickle.encode, documents)
     return docs_as_jsons
 
+
 def scrap_and_dump(destination_folder):
     """
     scrap and dump one document by file
-    :param destination_folder: folder must already exist
+    :param destination_folde_convert_to_utf8r: folder must already exist
```
Comment (nicoxx) [inline src/scraper/scraper/scraper.py]: oops, it's a typo, i will fix it

File: requirements.txt (hunk 2 above)
Comment (m-mokhtari) [inline requirements.txt]: Why urllib3 has been kept ?

File: src/scraper/scraper/scraper.py [hunk 5]
```diff
@@ -1,31 +1,52 @@
...
-    except (urllib2.URLError, ssl.SSLError): # when login/certificate is required to connect
+        data = requests.get(url, timeout=1)
+        header_encoding = data.encoding
+        if not header_encoding:
+            return None
+        data.content.decode(header_encoding.lower())  # check we can get unicode object from raw data (could raise exception)
+        guessed_encoding = data.apparent_encoding  # important to cache data.apparent_encoding only once (costly)
+        if guessed_encoding is None or header_encoding.lower() != guessed_encoding.lower():
+            return None
+        return data.text.encode('utf-8').decode('utf-8')  # get unicode object in utf-8 format
```
Comment (m-mokhtari) [inline src/scraper/scraper/scraper.py]: I don't understand this line.
1. Is data.text already an unicode object ("requests" library convert from data.encoding to unicode) ?
    a. If it's the case, is it necessary to do encode/decode ? Calling encode imply that data.text is already an unicode object ?
    b. If not, we should rather do: data.text.decode(data.encoding) ? To convert from binary data encoded in  
        data.encoding format to an unicode object ?

I'm missing something ?

File: src/scraper/tests/test_integration_scraper.py [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+                self.assertIsInstance(doc, Document)  # deserialization is ok
+                self.assertIsNotNone(doc.html_content)  # we don't return unavailable pages
+                self.assertIsInstance(doc.html_content, unicode)  # 'str' sucks, must use unicode (python 3 str versions)
+                self.assertNotIn('.gif', doc.link_element.url)  # check extension filter
+                self.assertNotIn('youtu', doc.link_element.url)  # check regex filter
+
+                curr_doc += 1
+                if curr_doc == nb_doc:
+                    break
+            else:
```
Comment (m-mokhtari) [inline src/scraper/tests/test_integration_scraper.py]: Indentation problem ?

File: requirements.txt (hunk 2 above)
Comment (nicoxx) [inline requirements.txt]: it still used, see reddit.py L14:
"# bug on me PC, see https://urllib3.readthedocs.org/en/latest/security.html section OpenSSL / PyOpenSSL"

File: src/scraper/scraper/scraper.py (hunk 5 above)
Comment (nicoxx) [inline src/scraper/scraper/scraper.py]: https://en.wikipedia.org/wiki/Cargo_cult_programming

it seems to work fine without it and I will remove it.

For the story, what I understood : data.text is an unicode object but the internal byte format is "data.encoding", not necessarily utf-8. The call to "encode" get me a raw byte (str) object, and "decode" will wrap it back to an unicode object, but with an utf-8 internal format.

File: src/scraper/tests/test_integration_scraper.py (hunk 6 above)
Comment (nicoxx) [inline src/scraper/tests/test_integration_scraper.py]: python powa
http://psung.blogspot.fr/2007/12/for-else-in-python.html

**Pull request merged**